adapters/src/format/mod.rs

```diff
@@ -0,0 +1,103 @@
+use crate::Catalog;
+use anyhow::Result as AnyResult;
+use once_cell::sync::Lazy;
+use serde_yaml::Value as YamlValue;
+use std::{
+    borrow::Cow,
+    collections::BTreeMap,
+    sync::{Arc, Mutex},
+};
+
+mod csv;
+
+use self::csv::CsvInputFormat;
+
+/// Static map of supported input formats.
+// TODO: support for registering new formats at runtime in order to allow
+// external crates to implement new formats.
+static INPUT_FORMATS: Lazy<BTreeMap<&'static str, Box<dyn InputFormat>>> =
+    Lazy::new(|| BTreeMap::from([("csv", Box::new(CsvInputFormat) as Box<dyn InputFormat>)]));
+
+/// Trait that represents a specific data format.
+///
+/// This is a factory trait that creates parsers for a specific data format.
+pub trait InputFormat: Send + Sync {
+    /// Unique name of the data format.
+    fn name(&self) -> Cow<'static, str>;
+
+    /// Create a new parser for the supported format.
+    ///
+    /// # Arguments
+    ///
+    /// * `config` - Format-specific configuration.
+    ///
+    /// * `catalog` - Circuit [catalog](`Catalog`).  The newly created parser
+    ///   will use this catalog to bind to one or more input streams of the
+    ///   circuit.  Most input endpoints feed data to one specific input stream.
+    ///   The name of this stream must be specified as an `input_stream` field
+    ///   of `config`.  Some endpoints, however, may carry data that gets
+    ///   demultiplexed to multiple input streams. An example is a database
+    ///   change data capture (CDC) protocol endpoint that receives updates to
+    ///   multiple input tables.  Its associated data format includes metadata
+    ///   about the destination table of each record.  The associated parser
+    ///   uses this metadata to locate corresponding steams in the circuit
+    ///   catalog.
+    fn new_parser(
+        &self,
+        config: &YamlValue,
+        catalog: &Arc<Mutex<Catalog>>,
+    ) -> AnyResult<Box<dyn Parser>>;
+}
+
+impl dyn InputFormat {
+    /// Lookup input format by name.
+    pub fn get_format(name: &str) -> Option<&'static dyn InputFormat> {
+        INPUT_FORMATS.get(name).map(|f| &**f)
```

Comment:
Doesn't like it:

```
error[E0308]: mismatched types
  --> adapters/src/format/mod.rs:55:9
   |
54 |     pub fn get_format(name: &str) -> Option<&'static dyn InputFormat> {
   |                                      -------------------------------- expected `std::option::Option<&'static (dyn InputFormat + 'static)>` because of return type
55 |         INPUT_FORMATS.get(name).as_deref()
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected trait object `dyn InputFormat`, found struct `Box`
   |
   = note: expected enum `std::option::Option<&'static (dyn InputFormat + 'static)>`
              found enum `std::option::Option<&Box<dyn InputFormat>>`
```